feat: sync AuthUserV1

shenlong-tanwen opened the pull request:
## Description

- Handles syncing `AuthUserV1` from the server
- Updates the current user provider / store to pull data from the auth user entity
- This might fix the incorrect quota usage displayed on the app since this PR updates it to be pulled from the auth user entity

shenlong-tanwen committed:
feat: sync AuthUserV1

shenlong-tanwen committed:
migration

shenlong-tanwen committed:
chore: fix analyze

mertalev commented on mobile/lib/domain/models/user.model.dart:
```diff
@@ -1,15 +1,44 @@
+  yellow("yellow"),
+  blue("blue"),
+  green("green"),
+  purple("purple"),
+  orange("orange"),
+  gray("gray"),
+  amber("amber");
+
+  final String value;
+  const AvatarColor(this.value);
+
+  Color toColor({bool isDarkTheme = false}) => switch (this) {
+    AvatarColor.primary => isDarkTheme ? const Color(0xFFABCBFA) : const Color(0xFF4250AF),
+    AvatarColor.pink => const Color.fromARGB(255, 244, 114, 182),
+    AvatarColor.red => const Color.fromARGB(255, 239, 68, 68),
+    AvatarColor.yellow => const Color.fromARGB(255, 234, 179, 8),
+    AvatarColor.blue => const Color.fromARGB(255, 59, 130, 246),
+    AvatarColor.green => const Color.fromARGB(255, 22, 163, 74),
+    AvatarColor.purple => const Color.fromARGB(255, 147, 51, 234),
+    AvatarColor.orange => const Color.fromARGB(255, 234, 88, 12),
+    AvatarColor.gray => const Color.fromARGB(255, 75, 85, 99),
+    AvatarColor.amber => const Color.fromARGB(255, 217, 119, 6),
+  };
+}
 
 // TODO: Rename to User once Isar is removed
 class UserDto {
   final String id;
   final String email;
   final String name;
   final bool isAdmin;
-  final DateTime updatedAt;
+  final DateTime? updatedAt;
```
I'm assuming this is nullable because it's used by both Isar and Drift and the latter doesn't have the `updatedAt` column? I'm not sure how to feel about this. The `updatedAt` being null or a default value would surely be problematic for the old sync with Isar, right?

mertalev commented on mobile/lib/domain/models/user.model.dart:
```diff
@@ -99,7 +128,7 @@ profileChangedAt: $profileChangedAt
     if (identical(this, other)) return true;
 
     return other.id == id &&
-        other.updatedAt.isAtSameMomentAs(updatedAt) &&
+        (updatedAt != null && other.updatedAt != null && other.updatedAt!.isAtSameMomentAs(updatedAt!)) &&
```
Is it intentional that two DTOs with null `updatedAt` will never be equal?

shenlong-tanwen replied:
Good catch! Fixed it

mertalev commented on mobile/lib/infrastructure/repositories/sync_stream.repository.dart:
```diff
@@ -29,6 +32,35 @@ class SyncStreamRepository extends DriftDatabaseRepository {
+        }
+      });
+    } catch (error, stack) {
+      _logger.severe('Error: SyncAuthUserV1', error, stack);
+      rethrow;
```
Is there a benefit to this? It seems like it'll just log the error twice.

shenlong-tanwen replied:
Not really, no. But I don't remember as to why we started doing it this way and not let it unwind back to the try catch in the api repo and log there

shenlong-tanwen committed:
fix user updatedAt check

alextran1502 commented on mobile/lib/services/sync.service.dart:
```diff
@@ -147,7 +147,9 @@ class SyncService {
       compare: (UserDto a, UserDto b) => a.id.compareTo(b.id),
       both: (UserDto a, UserDto b) {
-        if (!a.updatedAt.isAtSameMomentAs(b.updatedAt) ||
+        if ((a.updatedAt == null && b.updatedAt != null) ||
```
Why do we have more conditions to evaluate here?

alextran1502 reviewed (approved):
Looks good, what is the plan for `UserEntity`? 

https://github.com/immich-app/immich/blob/20a854733a99a5c86a1075740cb39b386ba02fdc/mobile/lib/infrastructure/entities/user.entity.dart#L78

jrasm91 committed:
fix: auth user sync query

alextran1502 committed:
generate sql

shenlong-tanwen committed:
merge main

# Conflicts:
#	mobile/drift_schemas/main/drift_schema_v9.json
#	mobile/lib/infrastructure/repositories/db.repository.dart
#	mobile/lib/infrastructure/repositories/db.repository.drift.dart
#	mobile/lib/infrastructure/repositories/db.repository.steps.dart
#	mobile/lib/infrastructure/repositories/sync_stream.repository.dart
#	mobile/lib/repositories/auth.repository.dart
#	mobile/test/drift/main/generated/schema_v9.dart

shenlong-tanwen committed:
bump schema version and update migration
